src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,168 @@
+            first = newNode;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        Node<T> item = getNodeByIndex(index);
+        if (item == first) {
+            addFirst(item, value);
+            return;
+        }
+        addBefore(item, value);
+        return;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> item = getNodeByIndex(index);
+        return item.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        MyLinkedList.Node<T> item = getNodeByIndex(index);
+        T oldValue = item.item;
+        item.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToDelete = getNodeByIndex(index);
+        T removeItemValue = nodeToDelete.item;
+        unlink(nodeToDelete);
+        return removeItemValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> item = first;
+        for (int i = 0; i < size; i++) {
+            if (object == item.item || item.item != null && item.item.equals(object)) {
+                unlink(item);
+                return true;
+            }
+            item = item.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        return true;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        checkIndex(index);
+        if (index < (size >> 1)) {
+            Node<T> pointer = first;
+            for (int i = 0; i < index; i++) {
+                pointer = pointer.next;
+            }
+            return pointer;
+        } else {
+            Node<T> pointer = last;
+            for (int i = size - 1; i > index; i--) {
+                pointer = pointer.prev;
+            }
+            return pointer;
+        }
+
+    }
+
+    private void addLast(Node<T> last, T value) {
+        Node<T> newNode = new Node<>(last, value, null);
+        last.next = newNode;
+        this.last = newNode;
+        size++;
+    }
+
+    private void addBefore(Node<T> item, T value) {
+        Node<T> newNode = new Node<>(item.prev, value, item);
+        newNode.prev.next = newNode;
+        item.prev = newNode;
+        size++;
+    }
+
+    private void addFirst(Node<T> first, T value) {
```

Comment:
```suggestion
    private void addFirst(T value) {
```
You have **first** class var